flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/preprocessing/StringIndexer.scala

```diff
@@ -0,0 +1,163 @@
+import org.apache.flink.ml.preprocessing.StringIndexer.HandleInvalid
+
+import scala.collection.immutable.Seq
+
+/**
+  * StringIndexer maps a DataSet[String] to a DataSet[(String,Int)] where each label is
+  * associated with an index.The indices are in [0,numLabels) and are ordered by label
+  * frequencies. The most frequent label has index 0.
+  *
+  * @example
+  * {{{
+  *                val trainingDS: DataSet[String] = env.fromCollection(data)
+  *                val transformer = StringIndexer().setHandleInvalid("skip")
+  *
+  *                transformer.fit(trainingDS)
+  *                val transformedDS = transformer.transform(trainingDS)
+  * }}}
+  *
+  *
+  * You can manage unseen labels using HandleInvalid parameter. If HandleInvalid is
+  * set to "skip" (see example),then each line containing an unseen label is skipped.
+  * Otherwise an exception is raised.
+  *
+  * =Parameters=
+  *
+  * -[[HandleInvalid]]: Define how to handle unseen labels: by default is "skip"
+  *
+  *
+  */
+class StringIndexer extends Transformer[StringIndexer] {
+
+  private[preprocessing] var metricsOption: Option[Map[String, Int]] = None
+
+
+  /**
+    * Set the value to handle unseen labels
+    * @param value set to "skip" if you want to filter line with unseen labels
+    * @return StringIndexer instance with HandleInvalid value
+    */
+  def setHandleInvalid(value: String): this.type ={
+    parameters.add( HandleInvalid, value )
+    this
+  }
+
+}
+
+object StringIndexer {
+
+  case object HandleInvalid extends Parameter[String] {
+    val defaultValue: Option[String] = Some( "skip" )
+  }
+
+  // ==================================== Factory methods ========================================
+
+  def apply(): StringIndexer ={
+    new StringIndexer( )
+  }
+
+  // ====================================== Operations ===========================================
+
+  /**
+    *  Trains [[StringIndexer]] by learning the count of each labels in the input DataSet.
+    *
+    * @return [[FitOperation]] training the [[StringIndexer]] on string labels
+    */
+  implicit def fitStringIndexer ={
+    new FitOperation[StringIndexer, String] {
+      def fit(instance: StringIndexer, fitParameters: ParameterMap,
+        input: DataSet[String]): Unit = {
+        val metrics = extractIndices( input )
+        instance.metricsOption = Some( metrics )
+      }
+    }
+  }
+
+  /**
+    * Count the frequency of each label, sort them in a decreasing order and assign an index
+    *
+    * @param input input Dataset containing labels
+    * @return a map that returns for each label (key) its index (value)
+    */
+  private def extractIndices(input: DataSet[String]): Map[String, Int] ={
+
+    implicit val resultTypeInformation = createTypeInformation[(String, Int)]
+
+    val mapper = input
+      .map( s => (s, 1) )
+      .groupBy( 0 )
+      .reduce( (a, b) => (a._1, a._2 + b._2) )
+      .collect( )
```

Comment:
`collect()` returns all results (via an accumulator) to the calling program. I'm not seeing why this would be necessary.